File: gfg-weekly-coding-contest-103/K-periodic Circular String.py
#User function Template for python3

class Solution:
    # Version 1: Cycle
    # s[i] = s[i + K] = s[i + 2K] = .. = s[i]
    # Count the length of cycle, and check whether the frequency of character can be divided by the length.
    # We can use gcd(n, K) to know the smallest step.
    # TC: O(n), SC: O(n)
    def kPeriodic(self, s, K): 
        #code here
        import math
        from collections import Counter
        n = len(s)
        fre = Counter(s)
        group_count = math.gcd(n, K)
        item_count = n // group_count
        for _, v in fre.items():
            if v % item_count > 0:
                return "-1"
        keys = sorted(fre.keys())
        ans = [""]*n
        i = 0
        for c in sorted(fre.keys()):
            while fre[c] > 0:
                for s in range(item_count):
                    ans[i + s * group_count] = c
                fre[c] -= item_count
                i += 1
        return ''.join(ans)

File: gfg-weekly-coding-contest-103/K-periodic Circular String.java


//User function Template for Java

class Solution 
{ 
    // Version 1: Cycle
    // s[i] = s[i + K] = s[i + 2K] = .. = s[i]
    // Count the length of cycle, and check whether the frequency of character can be divided by the length.
    // Put the smaller available alphabet in the fronter group.
    // TC: O(n), SC: O(n)
    String kPeriodic(String s, int K) 
    { 
        // code here
        int[] fre = new int[26];
        for (char c: s.toCharArray()) {
            fre[c - 'a'] += 1;
        }
        
        int n = s.length();
        ArrayList<ArrayList<Integer>> group = new ArrayList<>();
        int count = 0;
        for (int i = 0; count < n; i++) {
            int j = i;
            ArrayList<Integer> g = new ArrayList<>();
            g.add(j);
            j = (j + K) % n;
            while (j != i) {
                g.add(j);
                j = (j + K) % n;
            }
            group.add(g);
            count += g.size();
        }
        
        int groupSize = group.get(0).size();
        
        for (int i = 0; i < 26; i++) {
            if (fre[i] % groupSize > 0) {
                return "-1";
            }
        }
        
        char[] res = new char[n];
        int j = 0;
        for (int i = 0; i < 26; i++) {
            while (fre[i] > 0) {
                for (int index : group.get(j)) {
                    res[index] = (char)('a' + i);
                }
                fre[i] -= groupSize;
                j += 1;
            }
        }
        return new String(res);
        
    }
}